<!DOCTYPE html>
<html  lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

      <title>eco_helper.normalise</title>
    
          <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
          <link rel="stylesheet" href="_static/theme.css " type="text/css" />
      
      <!-- sphinx script_files -->
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>

      
      <!-- bundled in js (rollup iife) -->
      <!-- <script src="_static/theme-vendors.js"></script> -->
      <script src="_static/theme.js" defer></script>
    
  <link rel="index" title="Index" href="genindex.html" />
  <link rel="search" title="Search" href="search.html" />
  <link rel="next" title="eco_helper.format" href="eco_helper.format.html" />
  <link rel="prev" title="eco_helper.convert" href="eco_helper.convert.html" /> 
  </head>

  <body>
    <div id="app">
    <div class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="index.html" class="home-link">
    
      <img class="logo" src="_static/eco_helper.svg" alt="logo"/>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">



    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            



            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="index.html#welcome-to-eco-helper-s-documentation">Contents:</a></span>
      </p>
      <ul class="current">
        
          <li class="toctree-l1 ">
            
              <a href="eco_helper.convert.html" class="reference internal ">eco_helper.convert</a>
            

            
          </li>

        
          <li class="toctree-l1 current">
            
              <a href="#" class="reference internal current">eco_helper.normalise</a>
            

            
              <ul>
                
                  <li class="toctree-l2"><a href="#usage" class="reference internal">Usage</a></li>
                
                  <li class="toctree-l2"><a href="#full-cli" class="reference internal">Full CLI</a></li>
                
                  <li class="toctree-l2"><a href="#module-eco_helper.normalise.NormTable" class="reference internal">eco_helper.normalise.NormTable module</a></li>
                
                  <li class="toctree-l2"><a href="#module-eco_helper.normalise.funcs" class="reference internal">eco_helper.normalise.funcs module</a></li>
                
              </ul>
            
          </li>

        
          <li class="toctree-l1 ">
            
              <a href="eco_helper.format.html" class="reference internal ">eco_helper.format</a>
            

            
          </li>

        
          <li class="toctree-l1 ">
            
              <a href="eco_helper.core.html" class="reference internal ">eco_helper.core</a>
            

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
    
    <li>eco_helper.normalise</li>
  </ul>
  

  <ul class="page-nav">
  <li class="prev">
    <a href="eco_helper.convert.html"
       title="previous chapter">← eco_helper.convert</a>
  </li>
  <li class="next">
    <a href="eco_helper.format.html"
       title="next chapter">eco_helper.format →</a>
  </li>
</ul>
  
</div>
<hr>
          <div class="content" role="main" v-pre>
            
  <section id="module-eco_helper.normalise">
<span id="eco-helper-normalise"></span><span id="normalise"></span><h1>eco_helper.normalise<a class="headerlink" href="#module-eco_helper.normalise" title="Permalink to this heading">¶</a></h1>
<p>This module provides the functionality to normalise raw counts data to TPM or CPM.
It is the core of the <code class="docutils literal notranslate"><span class="pre">eco_helper</span> <span class="pre">normalise</span></code> command.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This subcommand exclusively works with <strong>TSV</strong> (tab-delimited) formatted files!</p>
</div>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this heading">¶</a></h2>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">eco_helper</span> <span class="n">normalise</span> <span class="o">&lt;</span><span class="n">norm</span><span class="o">&gt;</span> <span class="p">[</span><span class="o">--</span><span class="n">lengths</span> <span class="o">&lt;</span><span class="n">lengths</span><span class="o">&gt;</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gtf</span> <span class="o">&lt;</span><span class="n">gtf</span><span class="o">&gt;</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">names</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span> <span class="o">&lt;</span><span class="n">output</span><span class="o">&gt;</span><span class="p">]</span> <span class="o">&lt;</span><span class="nb">input</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">&lt;norm&gt;</span></code> is the kind of normalisation to perform, which can be eithe <code class="docutils literal notranslate"><span class="pre">tpm</span></code> or <code class="docutils literal notranslate"><span class="pre">cpm</span></code>, <code class="docutils literal notranslate"><span class="pre">&lt;input&gt;</span></code> is the input file, and <code class="docutils literal notranslate"><span class="pre">&lt;output&gt;</span></code> is the output file. By default the normalised data will be written to the same file as the input file.
In case of <code class="docutils literal notranslate"><span class="pre">tpm</span></code> also the lengths of the transcripts must be provided.
This can be done using the <code class="docutils literal notranslate"><span class="pre">--lengths</span></code> option.</p>
<p>It is possible to provide a <cite>GTF</cite> file to the <code class="docutils literal notranslate"><span class="pre">--gtf</span></code> option, instead of a lengths file.
In this case <cite>eco_helper</cite> will use <code class="docutils literal notranslate"><span class="pre">gtftools</span></code> to extract the lengths and use the <cite>merged</cite> transcript length.
Also, using the <code class="docutils literal notranslate"><span class="pre">--names</span></code> option the gene names (symbols) can be used instead of gene ids.</p>
</section>
<section id="full-cli">
<h2>Full CLI<a class="headerlink" href="#full-cli" title="Permalink to this heading">¶</a></h2>
<p>The full command line of <cite>eco_helper normalise</cite> with all options is as follows:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>usage: eco_helper normalise <span class="o">[</span>-h<span class="o">]</span> <span class="o">[</span>-o OUTPUT<span class="o">]</span> <span class="o">[</span>-l LENGTHS<span class="o">]</span> <span class="o">[</span>-g GTF<span class="o">]</span> <span class="o">[</span>-n<span class="o">]</span>
                        <span class="o">[</span>-d DIGITS<span class="o">]</span> <span class="o">[</span>-log<span class="o">]</span>
                        <span class="o">{</span>tpm,cpm<span class="o">}</span> input

Normalise raw counts data to TPM or CPM.

positional arguments:
<span class="o">{</span>tpm,cpm<span class="o">}</span>             The <span class="nb">type</span> of normalisation to perform. Can be either
                        <span class="s1">&#39;tpm&#39;</span> or <span class="s1">&#39;cpm&#39;</span>.
input                 Input file.

options:
-h, --help            show this <span class="nb">help</span> message and <span class="nb">exit</span>
-o OUTPUT, --output OUTPUT
                        Output file. By default the same as the input with
                        added suffix. The suffix will be either <span class="s1">&#39;.cpm&#39;</span> or
                        <span class="s1">&#39;.tpm&#39;</span>
-l LENGTHS, --lengths LENGTHS
                        Lengths file. If not provided, the lengths will be
                        extracted from the GTF file.
-g GTF, --gtf GTF     Reference GTF file <span class="k">for</span> transcript lengths and/or gene
                        names.
-n, --names           Use gene names instead of gene ids. This will replace
                        the gene ids <span class="o">(</span>index<span class="o">)</span> <span class="k">in</span> the expression matrix and
                        lengths file with gene symbols from the GTF file or
                        lengths file <span class="o">(</span><span class="k">if</span> provided<span class="o">)</span>. Note, <span class="k">if</span> a length file is
                        provided <span class="k">then</span> it must include gene names <span class="k">in</span> the second
                        column!
-d DIGITS, --digits DIGITS
                        The number of digits to round the values to.
-log, --logscale      Use this to log-scale the normalised values.
</pre></div>
</div>
<section id="using-gene-names">
<h3>Using Gene Names<a class="headerlink" href="#using-gene-names" title="Permalink to this heading">¶</a></h3>
<p>A note on using gene names instead of gene ids: By default <cite>eco_helper</cite> assumes that your input data uses Ensemble gene Ids as primary identifiers in the first column of your expression matrix.
Therefore, it will use the gene ids as index for extracted lengths and names form a given GTF file. If your data, however, works with gene names (symbols) inistead of ids, then no (tpm) normalisation will be possible
because <cite>eco_helper</cite> is trying to match lengths to genes using their index. If the data works on gene names and the lengths on gene ids then no overlap will be found! To prevent this, use the <code class="docutils literal notranslate"><span class="pre">--swap</span></code> when computing
new lengths to swap the gene ids (first column that is used for matching when normalising) with the gene names (second column). If you wish to use the second column for your normalised output file, then use the <code class="docutils literal notranslate"><span class="pre">--names</span></code> option.</p>
<blockquote>
<div><p>Examples:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">eco_helper</span> <span class="n">normalise</span> <span class="n">tpm</span> <span class="o">--</span><span class="n">swap</span> <span class="o">--</span><span class="n">gtf</span> <span class="n">gencode</span><span class="o">.</span><span class="n">GTF</span> <span class="o">--</span><span class="n">output</span> <span class="n">my_normalised</span><span class="o">.</span><span class="n">tsv</span> <span class="n">my_data</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<p>In the above example, <cite>my_data.tsv</cite> uses gene names (symbols) instead of gene ids. Therefore,
when computing new lengths from the GTF file we specify <code class="docutils literal notranslate"><span class="pre">--swap</span></code> to make sure the lengths file and our data
will match. However, <cite>my_output.tsv</cite> will use gene ids as index in the first column.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">eco_helper</span> <span class="n">normalise</span> <span class="n">tpm</span> <span class="o">--</span><span class="n">names</span> <span class="o">--</span><span class="n">gtf</span> <span class="n">gencode</span><span class="o">.</span><span class="n">GTF</span> <span class="o">--</span><span class="n">output</span> <span class="n">my_normalised</span><span class="o">.</span><span class="n">tsv</span> <span class="n">my_data</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
<p>In this second example, <cite>my_data.tsv</cite> uses gene ids as identifiers. Therefore, we do not need to swap the ids and names when computing the lengths.
However, <cite>my_output.tsv</cite> will use gene names as index in the first column instead of the gene ids because we specified <code class="docutils literal notranslate"><span class="pre">--names</span></code>.</p>
<p>The procedure also works backwards, of course. We can use <code class="docutils literal notranslate"><span class="pre">--swap</span></code> and <code class="docutils literal notranslate"><span class="pre">--names</span></code> to start with a data file using gene names and generate one that uses gene ids.</p>
</div></blockquote>
</section>
</section>
<section id="module-eco_helper.normalise.NormTable">
<span id="eco-helper-normalise-normtable-module"></span><h2>eco_helper.normalise.NormTable module<a class="headerlink" href="#module-eco_helper.normalise.NormTable" title="Permalink to this heading">¶</a></h2>
<p>A class to read raw counts data from an expression matrix file and normalise the data.
This is the main class of the <code class="docutils literal notranslate"><span class="pre">eco_helper</span> <span class="pre">normalise</span></code> command that will be called by the CLI.</p>
<dl class="py class">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">eco_helper.normalise.NormTable.</span></span><span class="sig-name descname"><span class="pre">NormTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class to read raw counts from an expression matrix file and normalise the data to either <code class="docutils literal notranslate"><span class="pre">tpm</span></code> or <code class="docutils literal notranslate"><span class="pre">cpm</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>str</em>) – The input count table. By default this is assumed to be tab-delimited.
Pass a <code class="docutils literal notranslate"><span class="pre">sep</span></code> argument to specify a different separator.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.adopt_name_index">
<span class="sig-name descname"><span class="pre">adopt_name_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.adopt_name_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Adopts the extracted name column of the lengths dataframe as the new
dataframe index for both the lengths and counts data.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This will only affect the raw and final counts (normalized),
but it will not affect the original counts!</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.counts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">counts</span></span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the raw or TPM counts (if normalisation was performed).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These are cropped to only genes that were found to have a
corresponding length in the lengths file and for which therefore
TPM conversion could be performed. If you wish to access the original
(pre filtered) data, use the raw_data attribute instead.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>counts</strong> – The raw or TPM counts.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the table of normalized values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The table.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.get_lengths">
<span class="sig-name descname"><span class="pre">get_lengths</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.get_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the lengths of the features.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lengths</strong> – The lengths of the features.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.Series or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ids</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ndarray</span></em><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the IDs of the features.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ids</strong> – The IDs of the features.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lengths</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ndarray</span></em><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the lengths of the features.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lengths</strong> – The lengths of the features.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy.ndarray or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.memorize">
<span class="sig-name descname"><span class="pre">memorize</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.memorize" title="Permalink to this definition">¶</a></dt>
<dd><p>Store the original (pre-filtered) and raw (unnormalised) counts.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.names">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">ndarray</span></em><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.names" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the names of the features.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>names</strong> – The names of the features.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.raw_counts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">raw_counts</span></span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.raw_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the raw counts.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These are cropped to only genes that were found to have a
corresponding length in the lengths file and for which therefore
TPM conversion could be performed. If you wish to access the original
(pre filtered) data, use the raw_data attribute instead.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>counts</strong> – The raw counts.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.raw_data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">raw_data</span></span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.raw_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the originally provided counts data.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This contains all the provided genes and their counts,
including those for which no lengths are available.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>raw_data</strong> – The originally provided counts data.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.read" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a table from a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – The input file.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – The separator of the table. The default is “        “.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – The table.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.round">
<span class="sig-name descname"><span class="pre">round</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">digits</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.round" title="Permalink to this definition">¶</a></dt>
<dd><p>Round tpm values to a given number of digits.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>digits</strong> (<em>int</em>) – The number of digits to round to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves the table to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – The output file.</p></li>
<li><p><strong>use_names</strong> (<em>bool</em>) – Save the file with gene_names instead of gene_ids in the first column.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.set_lengths">
<span class="sig-name descname"><span class="pre">set_lengths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.set_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the lengths of the features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – The file containing the lengths of the features.</p></li>
<li><p><strong>which</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name of the lengths. The default is None (in which case the last column is used).</p></li>
<li><p><strong>id_col</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name of the IDs. The default is None (in which case the first column is used).</p></li>
<li><p><strong>name_col</strong> (<em>str</em><em>, </em><em>optional</em>) – The column name of the (gene) names. The default is None (in which case the second column is used).
Note, even if your datafile does not specify gene names a “name column” will still be extracted.
However, you can adjust not to include the column later for saving the TPM-converted file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.to_cpm">
<span class="sig-name descname"><span class="pre">to_cpm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">digits</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.to_cpm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalise the raw counts to CPM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>digits</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of digits to round to. The default is 5.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the CPM values are logarithmically scaled. The default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="eco_helper.normalise.NormTable.NormTable.to_tpm">
<span class="sig-name descname"><span class="pre">to_tpm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">digits</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.NormTable.NormTable.to_tpm" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalise the raw counts to TPM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>digits</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of digits to round to. The default is 5.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the TPM values are logarithmically scaled. The default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-eco_helper.normalise.funcs">
<span id="eco-helper-normalise-funcs-module"></span><h2>eco_helper.normalise.funcs module<a class="headerlink" href="#module-eco_helper.normalise.funcs" title="Permalink to this heading">¶</a></h2>
<p>These are core functions of the normalise submodule.</p>
<dl class="py function">
<dt class="sig sig-object py" id="eco_helper.normalise.funcs.add_gtf_gene_names">
<span class="sig-prename descclassname"><span class="pre">eco_helper.normalise.funcs.</span></span><span class="sig-name descname"><span class="pre">add_gtf_gene_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">swap_ids_and_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.funcs.add_gtf_gene_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds the gene names to the GTF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – The input GTF file.</p></li>
<li><p><strong>outfile</strong> (<em>str</em>) – The output file.</p></li>
<li><p><strong>swap_ids_and_names</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to swap the IDs and names. The default is False.
If True then the Ids (1st column) and names (2nd column by default)
will be swapped so that names are the 1st column and IDs are the 2nd column.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="eco_helper.normalise.funcs.array_to_cpm">
<span class="sig-prename descclassname"><span class="pre">eco_helper.normalise.funcs.</span></span><span class="sig-name descname"><span class="pre">array_to_cpm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.funcs.array_to_cpm" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert raw counts to CPM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>np.ndarray</em>) – The raw counts. As a 2D ndarray.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use log-scale. The default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The CPM values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="eco_helper.normalise.funcs.array_to_tpm">
<span class="sig-prename descclassname"><span class="pre">eco_helper.normalise.funcs.</span></span><span class="sig-name descname"><span class="pre">array_to_tpm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lengths</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.funcs.array_to_tpm" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert raw counts to TPM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>array</strong> (<em>np.ndarray</em>) – The raw counts. As a 2D ndarray.</p></li>
<li><p><strong>lengths</strong> (<em>np.ndarray</em>) – The lengths of the features. As a 1D ndarray.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use log-scale. The default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The TPM values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="eco_helper.normalise.funcs.call_gtftools">
<span class="sig-prename descclassname"><span class="pre">eco_helper.normalise.funcs.</span></span><span class="sig-name descname"><span class="pre">call_gtftools</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'l'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.funcs.call_gtftools" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls gtftools from CLI to perform a computation.
By default to calculate lengths.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em>) – The input GTF file.</p></li>
<li><p><strong>output</strong> (<em>str</em>) – The output file.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – The mode of the computation. The default is “l”.
Any valid gtftools mode is allowed.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="eco_helper.normalise.funcs.match_regex_pattern">
<span class="sig-prename descclassname"><span class="pre">eco_helper.normalise.funcs.</span></span><span class="sig-name descname"><span class="pre">match_regex_pattern</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.funcs.match_regex_pattern" title="Permalink to this definition">¶</a></dt>
<dd><p>Matches a regex pattern to a dataframe using it’s “attributes” column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pattern</strong> (<em>str</em>) – The regex pattern.</p></li>
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – The dataframe.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The matched values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="eco_helper.normalise.funcs.round_values">
<span class="sig-prename descclassname"><span class="pre">eco_helper.normalise.funcs.</span></span><span class="sig-name descname"><span class="pre">round_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">digits</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#eco_helper.normalise.funcs.round_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Rounds an array of values to a certain number of digits.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>values</strong> (<em>np.ndarray</em>) – The raw values.</p></li>
<li><p><strong>digits</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of digits to round to. The default is 5.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The rounded values.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
  <li class="prev">
    <a href="eco_helper.convert.html"
       title="previous chapter">← eco_helper.convert</a>
  </li>
  <li class="next">
    <a href="eco_helper.format.html"
       title="next chapter">eco_helper.format →</a>
  </li>
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2022, Noah Kleinschmidt.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.1.1 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a> 0.8.0.
</div>
            </div>
          </div>
      </page>
    </div></div>
    
    
  </body>
</html>